components/apimgt/org.wso2.carbon.apimgt.usage/org.wso2.carbon.apimgt.usage.publisher/src/main/java/org/wso2/carbon/apimgt/usage/publisher/impl/AnalyticsDataPublisher.java

```diff
@@ -0,0 +1,86 @@
+/*
+ *  Copyright (c) 2021, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.wso2.carbon.apimgt.usage.publisher.impl;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.am.analytics.publisher.exception.MetricCreationException;
+import org.wso2.am.analytics.publisher.reporter.CounterMetric;
+import org.wso2.am.analytics.publisher.reporter.MetricReporter;
+import org.wso2.am.analytics.publisher.reporter.MetricReporterFactory;
+import org.wso2.am.analytics.publisher.reporter.MetricSchema;
+import org.wso2.carbon.apimgt.impl.APIManagerAnalyticsConfiguration;
+import org.wso2.carbon.apimgt.usage.publisher.Constants;
+import org.wso2.carbon.apimgt.usage.publisher.DataPublisherUtil;
+
+import java.util.Map;
+
+/**
+ * Analytics event publisher for APIM
+ */
+public class AnalyticsDataPublisher {
+    private static final Log log = LogFactory.getLog(AnalyticsDataPublisher.class);
+
+    private CounterMetric successMetricReporter;
+    private CounterMetric faultyMetricReporter;
+
+    private static AnalyticsDataPublisher instance;
+
+    private AnalyticsDataPublisher() {
+
+    }
+
+    public static synchronized AnalyticsDataPublisher getInstance() {
```

Comment:
Since this gets called frequently can we move the synchronised block into the method (after null check).

Reply:
yes +1

Reply:
Fixed